File: marg/src/marg/main/IMR.java
package marg.main;

import marg.exception.ParseException;
import marg.debug.Debug;
import marg.lexer.ILexer;
import marg.lexer.Lexer;
import marg.ast.ASTree;
import marg.lexer.SLexer;
import marg.parser.Environment;
import marg.parser.IParser;
import marg.parser.Parser;
import marg.token.Token;
import marg.token.TokenSet;
import marg.util.Extension;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.List;

import static marg.debug.Console.*;

import static org.fusesource.jansi.Ansi.*;
import static org.fusesource.jansi.Ansi.Color.*;

// Interactive Shell
public class IMR {

    public static void main(String[] args) {
        Debug.setEnabled(false);
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            out.print(ansi().reset().a('\n'));
            out.println("Program will exit...");
        }));

        Environment e = new Environment(null);
        SLexer lexer = new SLexer();
        IParser parser = new Parser();
        InputStreamReader isr = new InputStreamReader(System.in);
        BufferedReader reader = new BufferedReader(isr);

        try {
            while (true) {
                out.print("Marg> ");

                out.print(ansi().fg(GREEN));
                String s = reader.readLine();
                out.print(ansi().reset());

                if ("exit".equals(s)) {
                    System.exit(0);
                }
                if ("values".equals(s)) {
                    out.println("Values:");
                    e.map.entrySet().forEach(set ->
                            out.println(set.getKey() + " : " +set.getValue().get()));
                    continue;
                }
                if ("clear".equals(s)) {
                    out.println(ansi().eraseScreen());
                    continue;
                }

                List<Token> ls;
                try {
//                    ls = lexer.tokenize(s);
                    ls = Extension.convertList(lexer.tokenize(s));
                }
                catch (NullPointerException ex) {
                    // No input
                    continue;
                }

                ASTree ast;
                try {
                    ast = parser.parse(new TokenSet(ls));
                }
                catch (ParseException ex) {
                    out.println("Parse error.");
                    out.println(ex.getMessage());
                    ex.printStackTrace(out);
                    continue;
                }
                catch (Exception ex) {
                    out.println("Unexpected error.");
                    out.println(ansi().fg(RED).a(ex.getStackTrace()).reset());
                    continue;
                }

                try {
                    ast.eval(0, e);
                }
                catch (Exception ex) {
                    out.println("Runtime error.");
                    ex.printStackTrace(out);
                    continue;
                }
            }
        }
        catch (IOException ex) {
            ex.printStackTrace();
        }
    }
}

File: marg/src/marg/main/IMR.scala
package marg.main

import marg.exception.{ParseException, RuntimeError}
import marg.lexer.{ILexer, SLexer}
import marg.parser._
import marg.token.Token
import marg.util.AnsiExtension._

import scala.io.StdIn


object IMR {
  private var debug = false

  def main(args: Array[String]) {
    // Shutdown Hook
    sys addShutdownHook
      println("\n\n" + RESET + "Program will exit...")

    while (true) exec(new Env(null), new SLexer, new SParser)
  }


  def readLine(first: String, second: String): String = {
    print(if (first == null) second else first)
    print(GREEN)
    // FIXME
//    new jline.console.ConsoleReader().paste()
    val input = StdIn.readLine()
    print(RESET)

    if (!input.isEmpty && input.last == '\\')
      input.init + '\n' + readLine(null, second)
    else input
  }


  def exec(e: Env, lexer: ILexer, parser: IParser): Unit = {
    val input = readLine("Marg> ",
                         "    > ")

    if (input.isEmpty) return
    else if (input.head != ':') execLine(input, e, lexer, parser)
    else input.tail match {
      case "" =>
      case "?" | "h" | "help" => println(" <statement> : run <statement>")
      case "v" | "version" => println("Interactive Marg v0.1")
      case "q" | "quit" | "exit" => sys.exit()
      case "c" | "cls" | "clear" => print(CLEAR)
      case "e" | "env" | "environment" =>
        for (p <- e.map) println(s"${p._1} := ${p._2.get}")
      case "r" | "reset" => e.map.clear()
      case "d" | "debug" =>
        debug = !debug
        println(s"debug ${if (debug) "on" else "off"}.")

      case str => println(RED + s"${RED}Unknown command \'$str\'" + RESET)
    }

    println()
  }


  def execLine(input: String, e: Env, lexer: ILexer, parser: IParser) {
    val ls: List[Token] = try lexer.tokenize(input)
    catch {
      case ex: NullPointerException =>
        return

      case ex: Exception =>
        println("Failed to tokenize.")
        return
    }

    if (debug) println(YELLOW + ls.mkString("\n") + RESET)

    val ast = try parser.parse(ls)
    catch {
      case ex: ParseException =>
        println(RED + ex.getMessage + RESET)

        val offset = ex.getErrorOffset
        val list = ls.slice(offset - 3, offset + 3).init
        println(list.mkString("\n"))
        return

      case ex: Exception =>
        println("An unexpected error has occurred.")
        println(RED + ex.getStackTrace.slice(0, 10).mkString("\n ") + RESET)
        return
    }

    try {
      val eval = ast.eval(e)
      if (eval != null) println(eval.get)
    }
    catch {
      case ex: RuntimeError =>
        println("Runtime error!")
        println(BLUE + ex.getMessage + RESET)

      case ex: Exception =>
        println("An unexpected error has occurred while running.")
        println(YELLOW + ex.getClass + RESET)
        println(RED + ex.getStackTrace.slice(0, 10).mkString("\n ") + RESET)
        return
    }
  }
}
